### PR TITLE
Add back meta-device `assign=True` loading in merge_lora

### PR DESCRIPTION
**Pull request by carmocca**
This reverts commit d8dc97e4160307a674a6f4bd3ccdacbaa90b4c90.

### CODE REVIEW

File: litgpt/scripts/merge_lora.py [hunk 1]
```diff
@@ -43,16 +43,19 @@ def merge_lora(
...
-    with fabric.init_module():
+    with fabric.init_module(), torch.device("meta"):
         model = GPT(config)
 
     lora_path = checkpoint_dir / "lit_model.pth.lora"
     pretrained_checkpoint = torch.load(str(pretrained_checkpoint_dir / "lit_model.pth"), mmap=True)
     lora_checkpoint = torch.load(str(lora_path), mmap=True)
+    # since LoRA finetuning only saves the LoRA weights, we treat the lora weights dtype as the expected dtype
+    lora_dtype = next(iter(lora_checkpoint.values())).dtype
+    pretrained_checkpoint = {k: v.to(dtype=lora_dtype) for k, v in pretrained_checkpoint.items()}
```
Comment (awaelchli) [inline litgpt/scripts/merge_lora.py:54]: Can you point out what is actually now the difference of this vs how it is in main? Now that this conversion happens, mmap is not doing anything anymore. Same with assign: Before, `load_state_dict` would do the conversion and now we do it manually, what is the advantage? And there is actually no use for the meta device anymore now.

Comment (carmocca) [reply litgpt/scripts/merge_lora.py]: I had to do the conversion later

Comment (awaelchli) [reply litgpt/scripts/merge_lora.py]: Sure ok. In that sense, to answer my question, the main advantage is that we don't save the weights twice in memory, since the model's weights will be on the meta device. But other than the rest is still equivalent. Can you convirm?

Comment (carmocca) [reply litgpt/scripts/merge_lora.py]: I don't know what you mean by "the rest".

Yes, the entire point of assign is to not have two copies in memory, that's right.

The `.to()` call is to avoid the issue caused by us not saving the pretrained weights during fine-tuning.

Comment (awaelchli) [reply litgpt/scripts/merge_lora.py]: By the rest I mean the conversion step, which is done implicitly in main right now.
In a previous version, you did the conversion over the dictionary. So if this PR didn't save memory, naturally the question was what is the benefit? 
So in that sense, it doesn't revert what we've fixed in https://github.com/Lightning-AI/litgpt/commit/d8dc97e4160307a674a6f4bd3ccdacbaa90b4c90, but instead it reimplements #1189.

Comment (awaelchli) [reply litgpt/scripts/merge_lora.py]: Because the title of this PR is "Ensure pretrained and lora weights have the same dtype in merge_lora" which we already did in https://github.com/Lightning-AI/litgpt/commit/d8dc97e4160307a674a6f4bd3ccdacbaa90b4c90. That's what's confusing me.

**Comment by carmocca** [comment]
Blocked by #1385, #1378

**Comment by awaelchli** [comment]
@carmocca I lost a bit track of the intermediate changes that happened between my commit and this revert. But I'm good with merging if we're confident it was fixed properly. Ultimately, what needs to work is being able to merge a base model in float16 and a lora checkpoint in bfloat16, or vice versa. Which I think is covered in this test now: https://github.com/Lightning-AI/litgpt/pull/1251/files

**Review by awaelchli — APPROVED** [review]
